lib/rucio/tests/test_conveyor.py

```diff
@@ -757,6 +759,291 @@ def test_stager(rse_factory, did_factory, root_account, replica_client):
+    assert lock_core.get_replica_locks_for_rule_id(rule_id=rule_id)[0]['state'] == LockState.STUCK
+
+
+@skip_rse_tests_with_accounts
+@pytest.mark.noparallel(reason="runs submitter; poller and finisher")
+def test_transfer_to_mas_existing_replica(rse_factory, did_factory, root_account):
+    """
+    Test qos: transfer from tape to disk
+    Assert rse maximum_pin_lifetime is passed to transfer tool in the transfer request
+    Test rule and lock state transitions
+    """
+    src_rse, src_rse_id = rse_factory.make_rse(scheme='mock', protocol_impl='rucio.rse.protocols.posix.Default')
+    dst_rse, dst_rse_id = rse_factory.make_rse(scheme='mock', protocol_impl='rucio.rse.protocols.posix.Default', rse_type=RSEType.TAPE)
+    all_rses = [src_rse_id, dst_rse_id]
+
+    maximum_pin_lifetime = 86400
+
+    distance_core.add_distance(src_rse_id, dst_rse_id, ranking=10)
+    rse_core.add_rse_attribute(dst_rse_id, 'staging_required', True)
+    rse_core.add_rse_attribute(dst_rse_id, 'maximum_pin_lifetime', maximum_pin_lifetime)
+
+    did = did_factory.upload_test_file(rse_name=src_rse)
+
+    rule1_id = rule_core.add_rule(dids=[did], account=root_account, copies=1, rse_expression=dst_rse, grouping='ALL', weight=None, lifetime=-1, locked=False, subscription_id=None)[0]
+    request = request_core.get_request_by_did(rse_id=dst_rse_id, **did)
+
+    assert request['request_type'] == RequestType.TRANSFER
+
+    submitter(once=True, rses=[{'id': dst_rse_id}], partition_wait_time=None, transfertool='mock', transfertype='single', filter_transfertool=None)
+
+    assert request['attributes']['lifetime'] == str(maximum_pin_lifetime)
+    assert lock_core.get_replica_locks_for_rule_id(rule_id=rule1_id)[0]['state'] == LockState.REPLICATING
+    assert rule_core.get_rule(rule1_id)['state'] == RuleState.REPLICATING
+
+    # mock a successful transfer
+    request = request_core.get_request(request_id=request['id'])
+    request_core.set_request_state(request_id=request['id'], state=RequestState.DONE, external_id=request['external_id'])
+    finisher(once=True, partition_wait_time=None)
+
+    assert rule_core.get_rule(rule1_id)['state'] == RuleState.OK
+
+    # assert all replicas available
+    for rse_id in all_rses:
+        assert replica_core.get_replica(rse_id=rse_id, **did)['state'] == ReplicaState.AVAILABLE
+
+    # now test a second rule, different account
+    set_local_account_limit(InternalAccount('jdoe'), dst_rse_id, bytes_=-1)
+    rule2_id = rule_core.add_rule(dids=[did], account=InternalAccount('jdoe'), copies=1, rse_expression=dst_rse, grouping='ALL', weight=None, lifetime=-1, locked=False, subscription_id=None, source_replica_expression=dst_rse)[0]
+    request = request_core.get_request_by_did(rse_id=dst_rse_id, **did)
+
+    assert request['request_type'] == RequestType.STAGEIN
+
+    submitter(once=True, rses=[{'id': dst_rse_id}], partition_wait_time=None, transfertool='mock', transfertype='single', filter_transfertool=None)
```

Comment:
Technically you have to call the `stager` here. `submitter` only handles TRANSFER requests, stager handles STAGEIN. But it doesn't do anything anyway here since you have to mock everything. (Just for the sake of completeness)